In [1]:
!pip install textblob


# Exercise 1

In [2]:
from textblob import TextBlob

with open('Burbank.txt', 'r') as file: # reading the file
    burbank_text = file.read()

In [3]:
print(burbank_text[:200]) # checking the data

Airport task force to consider possible actions to abate noise issues Members of the Southern San Fernando Valley Airplane Noise Task Force will have a little over a month to develop recommendations 


In [4]:

blob = TextBlob(burbank_text) # creating a TextBlob object

polarity = blob.sentiment.polarity # getting the polarity of the text

print(polarity)


0.09869334480780263


##The polarity is 0.09869334480780263 i.e. it is positive
##hence, the probability of the sentiment going negative pretty low.

#Exercise 2

In [5]:
subjectivity = blob.sentiment.subjectivity # getting the subjectivity score

print("Polarity score:", polarity)
print("Subjectivity score:", subjectivity)

Polarity score: 0.09869334480780263
Subjectivity score: 0.3790877796901893


## With the polarity & subjectivity valuse being 0.09869334480780263 and 0.3790877796901893, we can say that the overall sentiment is positive and is more objective.


#Exercise 3

In [6]:
import nltk
nltk.download('punkt') # i had to download these to avoid errors
nltk.download('brown')
nltk.download('averaged_perceptron_tagger')
nltk.download('wordnet')

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package brown to /root/nltk_data...
[nltk_data]   Package brown is already up-to-date!
[nltk_data] Downloading package averaged_perceptron_tagger to
[nltk_data]     /root/nltk_data...
[nltk_data]   Package averaged_perceptron_tagger is already up-to-
[nltk_data]       date!
[nltk_data] Downloading package wordnet to /root/nltk_data...
[nltk_data]   Package wordnet is already up-to-date!


True

In [7]:
from textblob import Word
import string


topics = [] # extracting the nouns/proper nouns from the text
for wrd, tag in blob.tags: # using .tags functionality from textblob to get nouns
    if tag.startswith('NN'):
        cleaned_word = wrd.lower().strip(string.punctuation)  # removing the punctuations & cleaning text
        if cleaned_word.isalpha():
            topics.append(cleaned_word)


In [8]:
uniq_nouns = list(set(topics)) # getting the unique words/topics
print(uniq_nouns)

['task', 't', 'criteria', 'path', 'aviation', 'airplane', 'services', 'newcomer', 'system', 'los', 'quiet', 'altitude', 'procedures', 'hmmh', 'sectors', 'springer', 'resources', 'federal', 'radar', 'increase', 'vector', 'load', 'safety', 'llc', 'april', 'beautiful', 'relief', 'several', 'districts', 'force', 'training', 'waypoints', 'freeway', 'flight', 'public', 'transportation', 'president', 'jet', 'proposal', 'board', 'vice', 'routes', 'bulk', 'consulting', 'northbound', 'proposals', 'tracon', 'wednesday', 'community', 'karpe', 'news', 'terminal', 'weather', 'takeoff', 'hollywood', 'united', 'procedure', 'congress', 'people', 'uproarla', 'destinations', 'quicker', 'march', 'airfield', 'actions', 'attention', 'workload', 'way', 'andres', 'city', 'meetings', 'terms', 'noise', 'village', 'burbank', 'feb', 'senators', 'north', 'conduct', 'studio', 'experts', 'van', 'airlines', 'research', 'request', 'majority', 'ramirez', 'studies', 'east', 'quieter', 'angeles', 'recommendations', 'cont

In [9]:
tp_freqs = {} # a dictionary to store the topics & their frequencies

for noun in uniq_nouns:  # calculating & storing the frequencies using .words.count functionality from textblob
    count = blob.words.count(noun, case_sensitive=False)
    tp_freqs[noun] = count


sorted_tps = sorted(tp_freqs.items(), key=lambda x: x[1], reverse=True)[:15] # sorting the dictionary by frequency to get the top 10 frequent topics

print(sorted_tps)

[('faa', 13), ('burbank', 12), ('valley', 11), ('task', 10), ('force', 10), ('noise', 10), ('southern', 10), ('hmmh', 9), ('hollywood', 9), ('recommendations', 9), ('planes', 8), ('san', 8), ('flights', 8), ('flight', 7), ('groups', 7)]


In [10]:
print("These are the top 15 key topics")
for tp, freq in sorted_tps:
    print(f"Topic: {tp} and it's Frequency is: {freq}")

These are the top 15 key topics
Topic: faa and it's Frequency is: 13
Topic: burbank and it's Frequency is: 12
Topic: valley and it's Frequency is: 11
Topic: task and it's Frequency is: 10
Topic: force and it's Frequency is: 10
Topic: noise and it's Frequency is: 10
Topic: southern and it's Frequency is: 10
Topic: hmmh and it's Frequency is: 9
Topic: hollywood and it's Frequency is: 9
Topic: recommendations and it's Frequency is: 9
Topic: planes and it's Frequency is: 8
Topic: san and it's Frequency is: 8
Topic: flights and it's Frequency is: 8
Topic: flight and it's Frequency is: 7
Topic: groups and it's Frequency is: 7


#Exercise 4

In [11]:
import pandas as pd

In [12]:
data_amazon = pd.read_csv('amazon_cells_labelled.txt', delimiter='\t') # loading the datasets
data_imdb = pd.read_csv('imdb_labelled.txt', delimiter='\t')
data_yelp = pd.read_csv('yelp_labelled.txt', delimiter='\t')


print(data_amazon.head())

  So there is no way for me to plug it in here in the US unless I go by a converter.  \
0                        Good case, Excellent value.                                   
1                             Great for the jawbone.                                   
2  Tied to charger for conversations lasting more...                                   
3                                  The mic is great.                                   
4  I have to jiggle the plug to get it to line up...                                   

   0  
0  1  
1  1  
2  0  
3  1  
4  0  


In [13]:
data_amazon.columns = ['Review', 'Label'] # adding headers in each dataset
data_imdb.columns = ['Review', 'Label']
data_yelp.columns = ['Review', 'Label']

In [14]:
data_amazon['Company'] = 'Amazon' # adding a company column
data_imdb['Company'] = 'imdb'
data_yelp['Company'] = 'yelp'

In [15]:
combined_col = pd.concat([data_amazon, data_imdb, data_yelp], ignore_index=True) # combining the datasets

comb_data = combined_col # the new dataset

print(comb_data.head()) # checking the newdata

                                              Review  Label Company
0                        Good case, Excellent value.      1  Amazon
1                             Great for the jawbone.      1  Amazon
2  Tied to charger for conversations lasting more...      0  Amazon
3                                  The mic is great.      1  Amazon
4  I have to jiggle the plug to get it to line up...      0  Amazon


In [16]:
comb_data.to_csv('Sentiment_Analysis_Dataset.csv', index=False) # creating the Sentiment_Analysis_Dataset csv file


In [17]:
print(comb_data.columns) # cloumn names

Index(['Review', 'Label', 'Company'], dtype='object')


In [18]:
print(comb_data.isnull().sum()) # checking for nulls

Review     0
Label      0
Company    0
dtype: int64


In [19]:
import spacy

nlp = spacy.load('en_core_web_sm') # importing the STOP_WORDS from spacy and making a list of them
STOP_WORDS = nlp.Defaults.stop_words
stop_words = list(STOP_WORDS)

In [20]:
import string

punctuations = string.punctuation # getting the punctuations
print(punctuations)

def clean_text(txt_dt): # this is the basic function to clean the text

    cl_txt = ''.join([char for char in txt_dt if char not in punctuations])
    doc = nlp(cl_txt)  # creating a spaCy's parser object for cleaning & tokenizing
    tkns = []
    for tkn in doc: # filtering the text
        if tkn.text.lower() not in stop_words and tkn.text not in punctuations:
            tkns.append(tkn.text)
    return tkns

!"#$%&'()*+,-./:;<=>?@[\]^_`{|}~


In [21]:
from sklearn.base import TransformerMixin

class Predictors(TransformerMixin): # the prediction class
    def transform(self, X, **transform_params):
        return [' '.join(clean_text(text)) for text in X] # joining them back to strings

    def fit(self, X, y=None, **fit_params):
        return self

    def get_params(self, deep=True):
        return {}

In [22]:
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.model_selection import train_test_split


vectorizer = TfidfVectorizer(stop_words=stop_words) # using the TfidfVectorizer


X = comb_data['Review'] # defining the feature & target
y = comb_data['Label']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42) # splitting the data


In [23]:
print(X_train[:5])  # checking training data


2226    This place is a jewel in Las Vegas, and exactl...
1736    The opening sequence of this gem is a classic,...
1018    This if the first movie I've given a 10 to in ...
891                      Excellent product for the price.
2206    The vanilla ice cream was creamy and smooth wh...
Name: Review, dtype: object


In [24]:
from sklearn.pipeline import Pipeline
from sklearn.svm import LinearSVC
from sklearn.metrics import accuracy_score

pipe_countvect = Pipeline([ ('cleaner', Predictors()),('vectorizer', vectorizer),('classifier', LinearSVC())]) # creating the pipeline


pipe_countvect.fit(X_train, y_train) # training the model

preds = pipe_countvect.predict(X_test) # predicting with testing data


/usr/local/lib/python3.10/dist-packages/sklearn/feature_extraction/text.py:406: UserWarning: Your stop_words may be inconsistent with your preprocessing. Tokenizing the stop words generated tokens ['ll', 've'] not in stop_words.
  warnings.warn(


In [25]:
for i in range(20):
    print(f"{X_test.iloc[i]}, Prediction➔ {preds[i]}")


test_acc = accuracy_score(y_test, preds)
print("Accuracy with test data:", test_acc)

train_pred = pipe_countvect.predict(X_train) # predicting with training data
train_acc = accuracy_score(y_train, train_pred)
print("Accuracy with training data:", train_acc)


Disappointment.. I hate anything that goes in my ear., Prediction➔ 0
It is a true classic.  , Prediction➔ 1
Great product., Prediction➔ 1
This is a great restaurant at the Mandalay Bay., Prediction➔ 1
It finds my cell phone right away when I enter the car., Prediction➔ 1
It is simple to use and I like it., Prediction➔ 1
Of all the dishes, the salmon was the best, but all were great., Prediction➔ 1
I love the Pho and the spring rolls oh so yummy you have to try., Prediction➔ 1
Their Research and Development division obviously knows what they're doing., Prediction➔ 0
Still it's quite interesting and entertaining to follow.  , Prediction➔ 1
Very poor service., Prediction➔ 0
Oh yeah, and the storyline was pathetic too.  , Prediction➔ 0
Strike 2, who wants to be rushed., Prediction➔ 0
Every element of this story was so over the top, excessively phony and contrived that it was painful to sit through.  , Prediction➔ 0
The battery works great!, Prediction➔ 1
I am so tired of clichés that is ju